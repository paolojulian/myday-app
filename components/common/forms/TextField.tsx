import Label from '@/components/common/forms/Label';
import ThemedView from '@/components/common/ThemedView';
import { colors } from '@/constants/Colors';
import React, { ComponentProps, forwardRef, useState } from 'react';
import { StyleSheet, TextInput, TextInputProps } from 'react-native';

export type TextFieldProps = {
  label: string;
} & TextInputProps;

const TextField = forwardRef<TextInput, TextFieldProps>(({ label, style, ...props }, ref) => {
  const [isFocused, setIsFocused] = useState(false);

  const handleFocus: ComponentProps<typeof TextInput>['onFocus'] = e => {
    setIsFocused(true);
    if (props.onFocus) {
      props.onFocus(e);
    }
  };

  const handleBlur: ComponentProps<typeof TextInput>['onBlur'] = e => {
    setIsFocused(false);
    if (props.onBlur) {
      props.onBlur(e);
    }
  };

  return (
    <ThemedView style={[styles.container]}>
      <ThemedView style={styles.label}>
        <Label text={label} />
      </ThemedView>
      <TextInput
        {...props}
        ref={ref}
        placeholderTextColor={colors.grey}
        style={[
          style,
          styles.textInput,
          {
            ...(isFocused && styles.textInputWithFocus),
          },
        ]}
        onFocus={handleFocus}
        onBlur={handleBlur}
      />
    </ThemedView>
  );
});

TextField.displayName = 'TextField';

const styles = StyleSheet.create({
  container: {
    position: 'relative',
  },
  label: {
    zIndex: 1,
    position: 'absolute',
    top: 12,
    left: 24,
  },
  textInput: {
    fontSize: 20,
    paddingTop: 36,
    paddingBottom: 12,
    paddingHorizontal: 24,
    backgroundColor: colors.whiteSmoke,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: 'transparent',
  },
  textInputWithFocus: {
    borderColor: colors.black,
  },
});

export default TextField;
