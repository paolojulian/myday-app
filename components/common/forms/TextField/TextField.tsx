import Label from '@/components/common/forms/Label';
import ThemedText, { themedTextStyles } from '@/components/common/ThemedText/ThemedText';
import ThemedView from '@/components/common/ThemedView';
import { colors } from '@/constants/Colors';
import React, { ComponentProps, forwardRef, useMemo, useState } from 'react';
import { StyleSheet, TextInput, TextInputProps } from 'react-native';

export const TEXT_FIELD_TEST_IDS = {
  textInput: 'textField__textInput',
  errorMessage: 'textField__errorMessage',
};

export type TextFieldProps = {
  label: string;
  isError?: boolean;
  errorMessage?: string;
} & TextInputProps;

const TextField = forwardRef<TextInput, TextFieldProps>(
  ({ label, isError = false, errorMessage, style, ...props }, ref) => {
    const [isFocused, setIsFocused] = useState(false);

    const resolvedTextInputStyle = useMemo(() => {
      if (isError) {
        // Error should be the priority
        return styles.textInputWithError;
      }

      if (isFocused) {
        return styles.textInputWithFocus;
      }

      return styles.textInputDefault;
    }, [isFocused, isError]);

    const handleFocus: ComponentProps<typeof TextInput>['onFocus'] = e => {
      setIsFocused(true);
      if (props.onFocus) {
        props.onFocus(e);
      }
    };

    const handleBlur: ComponentProps<typeof TextInput>['onBlur'] = e => {
      setIsFocused(false);
      if (props.onBlur) {
        props.onBlur(e);
      }
    };

    return (
      <ThemedView style={[styles.container]}>
        <ThemedView style={styles.label}>
          <Label text={label} />
        </ThemedView>
        <TextInput
          {...props}
          ref={ref}
          testID={TEXT_FIELD_TEST_IDS.textInput}
          placeholderTextColor={colors.grey}
          style={[style, styles.textInput, themedTextStyles.body1, resolvedTextInputStyle]}
          onFocus={handleFocus}
          onBlur={handleBlur}
        />

        {!!errorMessage && (
          <ThemedText testID={TEXT_FIELD_TEST_IDS.errorMessage} style={{ color: colors.red }}>
            {errorMessage}
          </ThemedText>
        )}
      </ThemedView>
    );
  },
);

TextField.displayName = 'TextField';

const styles = StyleSheet.create({
  container: {
    position: 'relative',
  },
  label: {
    zIndex: 1,
    position: 'absolute',
    top: 12,
    left: 16,
  },
  textInput: {
    paddingTop: 36,
    paddingBottom: 12,
    paddingHorizontal: 16,
    backgroundColor: colors.whiteSmoke,
    borderRadius: 8,
    borderWidth: 1,
  },
  textInputDefault: {
    borderColor: 'transparent',
  },
  textInputWithFocus: {
    borderColor: colors.black,
  },
  textInputWithError: {
    borderColor: colors.red,
  },
});

export default TextField;
