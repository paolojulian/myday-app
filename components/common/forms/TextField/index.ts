export { default, type TextFieldProps } from './TextField';
