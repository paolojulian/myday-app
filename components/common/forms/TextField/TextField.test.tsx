import TextField, { TEXT_FIELD_TEST_IDS } from '@/components/common/forms/TextField/TextField';
import { act, render, screen } from '@testing-library/react-native';

describe('TESTING TextField component', () => {
  describe('WHEN the TextField component is rendered', () => {
    it('THEN it should match the snapshot', () => {
      const component = render(
        <TextField
          onChangeText={jest.fn()}
          label="Title"
          value="test"
          placeholder="Search field"
        />,
      );
      expect(component).toMatchSnapshot();
    });
    it('THEN it should not crash', () => {
      const component = render(
        <TextField
          onChangeText={jest.fn()}
          label="Title"
          value="test"
          placeholder="Search field"
        />,
      );
      expect(component).toBeTruthy();
    });
  });

  describe('WHEN the TextField is focused', () => {
    it('THEN it should call the onFocus prop', () => {
      const spyFocus = jest.fn();
      render(
        <TextField
          onChangeText={jest.fn()}
          onFocus={spyFocus}
          label="Title"
          value="test"
          placeholder="Search field"
        />,
      );
      const textInput = screen.getByTestId(TEXT_FIELD_TEST_IDS.textInput);
      expect(textInput).toBeTruthy();
      act(() => {
        textInput.props.onFocus();
      });

      expect(spyFocus).toHaveBeenCalled();
    });
  });

  describe('WHEN the TextField has errors', () => {
    it('THEN it should match the snapshot', () => {
      const component = render(
        <TextField
          onChangeText={jest.fn()}
          label="Title"
          value=""
          placeholder="Search field"
          isError
          errorMessage="Title is required."
        />,
      );
      expect(component).toMatchSnapshot();
    });
    it('THEN it should display the error message', () => {
      const spyFocus = jest.fn();
      render(
        <TextField
          onChangeText={jest.fn()}
          onFocus={spyFocus}
          label="Title"
          value=""
          placeholder="Search field"
          isError
          errorMessage="Title is required."
        />,
      );
      const errorMessage = screen.getByTestId(TEXT_FIELD_TEST_IDS.errorMessage);
      expect(errorMessage).toBeTruthy();
      expect(errorMessage.children).toHaveLength(1);
      expect(errorMessage.children[0]).toBe('Title is required.');
    });
  });
});
