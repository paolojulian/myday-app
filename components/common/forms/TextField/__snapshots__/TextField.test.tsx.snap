// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TESTING TextField component WHEN the TextField component is rendered THEN it should match the snapshot 1`] = `
<View
  style={
    [
      {
        "position": "relative",
      },
    ]
  }
>
  <View
    style={
      {
        "left": 16,
        "position": "absolute",
        "top": 12,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "PoppinsMedium",
            "fontSize": 16,
            "fontWeight": 500,
          },
          {
            "fontWeight": 400,
          },
          {
            "color": "#0F172A",
          },
        ]
      }
    >
      Title
    </Text>
  </View>
  <TextInput
    onBlur={[Function]}
    onChangeText={[MockFunction]}
    onFocus={[Function]}
    placeholder="Search field"
    placeholderTextColor="#a3a3a3"
    style={
      [
        undefined,
        {
          "backgroundColor": "#f5f5f5",
          "borderRadius": 8,
          "borderWidth": 1,
          "paddingBottom": 12,
          "paddingHorizontal": 16,
          "paddingTop": 36,
        },
        {
          "fontFamily": "PoppinsRegular",
          "fontSize": 16,
        },
        {
          "borderColor": "transparent",
        },
      ]
    }
    testID="textField__textInput"
    value="test"
  />
</View>
`;

exports[`TESTING TextField component WHEN the TextField has errors THEN it should match the snapshot 1`] = `
<View
  style={
    [
      {
        "position": "relative",
      },
    ]
  }
>
  <View
    style={
      {
        "left": 16,
        "position": "absolute",
        "top": 12,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "PoppinsMedium",
            "fontSize": 16,
            "fontWeight": 500,
          },
          {
            "fontWeight": 400,
          },
          {
            "color": "#0F172A",
          },
        ]
      }
    >
      Title
    </Text>
  </View>
  <TextInput
    onBlur={[Function]}
    onChangeText={[MockFunction]}
    onFocus={[Function]}
    placeholder="Search field"
    placeholderTextColor="#a3a3a3"
    style={
      [
        undefined,
        {
          "backgroundColor": "#f5f5f5",
          "borderRadius": 8,
          "borderWidth": 1,
          "paddingBottom": 12,
          "paddingHorizontal": 16,
          "paddingTop": 36,
        },
        {
          "fontFamily": "PoppinsRegular",
          "fontSize": 16,
        },
        {
          "borderColor": "#CC3333",
        },
      ]
    }
    testID="textField__textInput"
    value=""
  />
  <Text
    style={
      [
        {
          "fontFamily": "PoppinsRegular",
          "fontSize": 14,
        },
        {
          "fontWeight": 400,
        },
        {
          "color": "#CC3333",
          "paddingTop": 8,
        },
      ]
    }
    testID="textField__errorMessage"
  >
    Title is required.
  </Text>
</View>
`;
